sdk/servicebus/azure-servicebus/azure/servicebus/_common/auto_lock_renewer.py

```diff
@@ -0,0 +1,132 @@
+        _log.debug("Running lock auto-renew thread for %r seconds", timeout)
+        error = None
+        try:
+            while self._renewable(renewable):
+                if (utc_now() - starttime) >= datetime.timedelta(seconds=timeout):
+                    _log.debug("Reached auto lock renew timeout - letting lock expire.")
+                    raise AutoLockRenewTimeout("Auto-renew period ({} seconds) elapsed.".format(timeout))
+                if (renewable.locked_until_utc - utc_now()) <= datetime.timedelta(seconds=self.renew_period):
+                    _log.debug("%r seconds or less until lock expires - auto renewing.", self.renew_period)
+                    renewable.renew_lock()
+                time.sleep(self.sleep_time)
+        except AutoLockRenewTimeout as e:
+            renewable.auto_renew_error = e
+        except Exception as e:  # pylint: disable=broad-except
+            _log.debug("Failed to auto-renew lock: %r. Closing thread.", e)
+            error = AutoLockRenewFailed(
+                "Failed to auto-renew lock",
+                inner_exception=e)
+            renewable.auto_renew_error = error
+        finally:
+            if on_lock_renew_failure:
+                if self._shutdown.is_set() \
+                        or (hasattr(renewable, 'settled') and renewable.settled) \
+                        or (isinstance(renewable, ServiceBusSession) and not renewable._receiver._running) \
+                        or (not error and not renewable.expired):
+                    # Basically we want to make sure that any of the "intentional" exit cases are not fired on.
+                    # e.g. shutdown, settlement, session closure, renewer timeout (as opposed to renew failure/expiry)
+                    return
+                on_lock_renew_failure(renewable)
+
+    def register(self, renewable, timeout=300, on_lock_renew_failure=None):
+        """Register a renewable entity for automatic lock renewal.
+
+        :param renewable: A locked entity that needs to be renewed.
+        :type renewable: ~azure.servicebus.ReceivedMessage or
+         ~azure.servicebus.Session
+        :param float timeout: A time in seconds that the lock should be maintained for.
+         Default value is 300 (5 minutes).
+        :param Callable[[Union[Session,ReceivedMessage]], None] on_lock_renew_failure: A callback may be specified
+         to be called when the lock is lost on the renewable that is being registered.
+         Default value is None (no callback).
+        """
+        if self._shutdown.is_set():
+            raise ServiceBusError("The AutoLockRenew has already been shutdown. Please create a new instance for"
+                                  " auto lock renewing.")
+        starttime = renewable_start_time(renewable)
+        self.executor.submit(self._auto_lock_renew, renewable, starttime, timeout, on_lock_renew_failure)
+
+    def shutdown(self, wait=True):
```

Comment:
this might be unrelated topic, I just want to make sure it doesn't fall off our track.

Do we want to rename "shutdown" to "close"?
I remember in the user-study, we found naming consistency matters.

Reply:
Hasn't fallen off :) (It's in the list of remaining UX study action items that haven't been marked complete)

Apt reminder though; I just wanted to discuss it with @annatisch beforehand because there's technically a semantic difference between close()ing a client (with a "connection" inside it) and "shutting down" a coroutine, and I was not sure if that was something we had done intentionally.

Reply:
Close was settled upon++